### PR TITLE
allow to extract values from array of objects with jsonPathArray

### PR DESCRIPTION
**Pull request by klsince**
Allow to extract values from array of objects with jsonPathArray. 
And add jsonPathArrayDefaultEmpty to allow skip json records without target path, instead of aborting the ingestion.

## Upgrade Notes
Does this PR prevent a zero down-time upgrade? (Assume upgrade order: Controller, Broker, Server, Minion)
* [ ] Yes (Please label as **<code>backward-incompat</code>**, and complete the section below on Release Notes)

Does this PR fix a zero-downtime upgrade introduced earlier?
* [ ] Yes (Please label this as **<code>backward-incompat</code>**, and complete the section below on Release Notes)

Does this PR otherwise need attention when creating release notes? Things to consider:
- New configuration options
- Deprecation of configurations
- Signature changes to public methods/interfaces
- New plugins added or old plugins removed
* [ ] Yes (Please label this PR as **<code>release-notes</code>** and complete the section on Release Notes)

### CODE REVIEW

**Comment by klsince** [comment]
As context: 

JsonPath assumes json array is deserialized into a Java Collection ([code snippet](https://github.com/json-path/JsonPath/blob/master/json-path/src/main/java/com/jayway/jsonpath/spi/json/AbstractJsonProvider.java#L32)). But Pinot explicitly converts Collection to Object[] ([code snippet](https://github.com/apache/pinot/blob/master/pinot-spi/src/main/java/org/apache/pinot/spi/data/readers/BaseRecordExtractor.java#L47)) before calling those json transformations, thus leading to issues like: either no values were extracted with ingestion transformation like `$.commits[*].sha` or throwing PathNotFoundException with sth like `$.commits[0].sha`. 

The logic to covert Collection to Object[] is necessary, as Object[] is assumed by some transformations, used for MV values on both ingestion and query path. Thus this RP is trying to extend JsonPath to be able to work with Object[] properly.

File: pinot-common/src/main/java/org/apache/pinot/common/function/scalar/JsonFunctions.java [hunk 1]
```diff
@@ -109,12 +110,19 @@ public static Object jsonPath(Object object, String jsonPath) {
...
     }
-    if (object instanceof Object[]) {
-      return convertObjectToArray(JsonPath.read(JsonUtils.objectToString(object), jsonPath));
-    }
     return convertObjectToArray(JsonPath.read(object, jsonPath));
   }
 
+  @ScalarFunction
+  public static Object[] jsonPathArrayDefaultEmpty(Object object, String jsonPath)
+      throws JsonProcessingException {
```
Comment (Jackie-Jiang) [inline pinot-common/src/main/java/org/apache/pinot/common/function/scalar/JsonFunctions.java:118]: It won't throw any exception

File: pinot-common/src/main/java/org/apache/pinot/common/function/scalar/JsonFunctions.java [hunk 2]
```diff
@@ -109,12 +110,19 @@ public static Object jsonPath(Object object, String jsonPath) {
...
+    try {
+      return jsonPathArray(object, jsonPath);
+    } catch (Exception e) {
+      return new Object[]{};
```
Comment (Jackie-Jiang) [inline pinot-common/src/main/java/org/apache/pinot/common/function/scalar/JsonFunctions.java:122]: (nit)
```suggestion
      return new Object[0];
```

File: pinot-common/src/main/java/org/apache/pinot/common/function/scalar/JsonFunctions.java [hunk 3]
```diff
@@ -199,4 +207,47 @@ public static double jsonPathDouble(Object object, String jsonPath, double defau
       return defaultValue;
     }
   }
+
+  static class ArrayAwareJacksonJsonProvider extends JacksonJsonProvider {
```
Comment (Jackie-Jiang) [inline pinot-common/src/main/java/org/apache/pinot/common/function/scalar/JsonFunctions.java:211]: Annotate with `VisibleForTesting`

File: pinot-common/src/main/java/org/apache/pinot/common/function/scalar/JsonFunctions.java [hunk 4]
```diff
@@ -199,4 +207,47 @@ public static double jsonPathDouble(Object object, String jsonPath, double defau
...
+    @Override
+    public Object getArrayIndex(Object obj, int idx) {
+      if (obj instanceof Object[]) {
+        return ((Object[]) obj)[idx];
+      }
+      return super.getArrayIndex(obj, idx);
+    }
+
+    @Override
+    public void setArrayIndex(Object obj, int index, Object newValue) {
```
Comment (Jackie-Jiang) [inline pinot-common/src/main/java/org/apache/pinot/common/function/scalar/JsonFunctions.java:226]: This is probably not required?

**Review by Jackie-Jiang — APPROVED** [review]
LGTM with minor comments

File: pinot-common/src/main/java/org/apache/pinot/common/function/scalar/JsonFunctions.java (hunk 1 above)
Comment (klsince) [reply pinot-common/src/main/java/org/apache/pinot/common/function/scalar/JsonFunctions.java:118]: good catch. removed.

File: pinot-common/src/main/java/org/apache/pinot/common/function/scalar/JsonFunctions.java (hunk 4 above)
Comment (klsince) [reply pinot-common/src/main/java/org/apache/pinot/common/function/scalar/JsonFunctions.java:226]: looks like it's mainly for updating JSON object, like adding or setting a value at certain json path. Don't think it's needed for here, removing for now.

**Pull request merged**